### PR TITLE
Fixes #34323 - use defaults from memory

### PR DESCRIPTION
**Pull request by ezr-ondrej**
This loads values from database only if the values are in the database.
If there is no value populated for the Setting, we want to read the default from memory.

### CODE REVIEW

File: app/services/setting_registry.rb [hunk 1]
```diff
@@ -151,6 +151,7 @@ def known_categories
   def load_values(ignore_cache: false)
     # we are loading only known STIs as we load settings fairly early the first time and plugin classes might not be loaded yet.
     settings = Setting.unscoped.where(category: known_categories)
+    settings = settings.where.not(value: nil)
```
Comment (ekohl) [inline app/services/setting_registry.rb]: Can't these 2 be combined on the same line or even in the same where?

**Review by ekohl — COMMENTED** [review]
Trying to process this mentally. I wonder if there are cases where we want to override a default and make it `nil`. Is that something we do?

File: app/services/setting_registry.rb (hunk 1 above)
Comment (ezr-ondrej) [reply app/services/setting_registry.rb]: one is negated, so it can't be the same where, but it can be on the same line if you feel it would be better :)

File: config/settings.yaml.test [hunk 2]
```diff
@@ -1,5 +1,5 @@
 ---
 :unattended: true
 :require_ssl: false
-:fqdn: foreman.example.com
+:fqdn: foreman.some.host.fqdn
```
Comment (evgeni) [inline config/settings.yaml.test]: why is this change needed? I don't see it reflected in the tests

also, please use `example.test` or `example.invalid`, not a randomly invented `.fqdn` tld ;)

Comment (ezr-ondrej) [reply config/settings.yaml.test]: Well... this is the value that the default setting got from fixtures and until this point, we did not reflect the `fqdn` value, because test were overriding it with fixture. So I chose to set that value to what tests expect instead of changing all the tests, but I can change the tests instead :)

Comment (ekohl) [reply config/settings.yaml.test]: Then IMHO the fixtures should be changed. https://community.theforeman.org/t/using-ips-and-domains-in-tests-documentation/16600 should be leading here. Historically people have not respected it, but I think we should.

File: app/services/setting_registry.rb [hunk 3]
```diff
@@ -58,7 +58,7 @@ def [](name)
       logger.warn("Setting #{name} has no definition, please define it before using") if ready?
       return
     end
-    definition.value
+    definition.value || definition.default
```
Comment (ekohl) [inline app/services/setting_registry.rb]: If `definition.value` is `false` it'll also use `definition.default`. This must check for `nil?`.

Comment (ezr-ondrej) [reply app/services/setting_registry.rb]: if I would read comments sooner then debugging, I could have saved myself a little bit of struggle, you are correct and I've fixed it :+1:

**Comment by ezr-ondrej** [comment]
> Trying to process this mentally. I wonder if there are cases where we want to override a default and make it `nil`. Is that something we do?

Well actually yes, but then it's yaml encoded nil and I follow:
https://github.com/theforeman/foreman/blob/7364916bd25512c2065892fab96c439648dbb6f1/app/models/setting.rb#L124

So in case it's user defined nil, it should be yaml encoded and not `NULL` so we should be fine.

**Review by ekohl — COMMENTED** [review]
Given the size of the `foreman.some.host.fqdn` -> `foreman.example.com` change, would it make sense to make it a separate commit?

**Comment by ezr-ondrej** [comment]
> Given the size of the `foreman.some.host.fqdn` -> `foreman.example.com` change, would it make sense to make it a separate commit?

Yeah I was thinking about it already as well. I'll separate all the fixture sanitization once I get all the tests to pass.

**Comment by evgeni** [comment]
the katello failure will require changes in katello: https://github.com/Katello/katello/blob/068058eb0e85521c86299acb1183467f96c1ce3e/test/mailers/errata_mailer_test.rb#L41

File: app/services/setting_registry.rb [hunk 4]
```diff
@@ -182,7 +182,7 @@ def _add(name, category:, type:, default:, description:, full_name:, context:, v
 
   def _find_or_new_db_record(name)
     definition = find(name)
-    Setting.find_by(name: name) || _new_db_record(definition)
+    Setting.find_by(name: name.to_s) || _new_db_record(definition)
```
Comment (ekohl) [inline app/services/setting_registry.rb]: Could you help my understanding? Why is this needed?

**Comment by ezr-ondrej** [comment]
So I've splitted the fixtures sanitization in separate commit and splitted the fqdn sanitization in a separate PR: #9068

File: app/services/setting_registry.rb (hunk 4 above)
Comment (ezr-ondrej) [reply app/services/setting_registry.rb]: It is not, while debugging I was trying everything I could think of and this is a left over (that didn't help ofc) :)
I can get rid of this, but thought it's safer to always have a string just as a safety precaution :shrug:

Comment (ezr-ondrej) [reply app/services/setting_registry.rb]: Reverted as you suggested, it might confuse future readers why we did this together in this patch :)

**Comment by ezr-ondrej** [comment]
:green_apple: If someone might give it a go, it would be awesome, it should get cherry-picked to 3.1 for sure.

**Comment by evgeni** [comment]
after applying this patch to my 3.1.1 system, my settings are empty?!

```console
# hammer setting info --name foreman_url
Id:            foreman_url
Name:          foreman_url
Description:   URL where your Foreman instance is reachable (see also Provisioning > unattended_url)
Category:      General
Settings type: string
Value:

```

It's correct in memory:

```ruby
irb(main):001:0> Setting[:foreman_url]
=> "https://sat70.tanso.example.com"
```

But:

```ruby
irb(main):009:0> Setting.find_by(name: 'foreman_url')
=> #<Setting id: 113, name: "foreman_url", value: nil, description: "URL where your Foreman instance is reachable (see ...", category: "Setting", settings_type: nil, default: "--- https://sat-7-0-qa-rhel7.tanso.example.com\n......", created_at: "2022-01-07 07:47:18", updated_at: "2022-01-07 07:47:18", full_name: nil, encrypted: false>
```

**Comment by ezr-ondrej** [comment]
@evgeni that doesn't confuse me at all. It just didn't work ^_^ - we were showing only the explicitly set values in API, which you do not have I suspect and that means it shows empty value (btw wouldn't it be helpful to see the explicitly set and effective value somewhere in the API)?
What's quite confusing is that none of the tests failed (meaning we did not validate the show response in API at all) :facepalm:

File: app/presenters/setting_presenter.rb [hunk 5]
```diff
@@ -51,7 +71,7 @@ def readonly?
   end
 
   def value
-    SETTINGS.fetch(name.to_sym) { super }
+    SETTINGS.fetch(name.to_sym) { explicit_value? ? super : default }
```
Comment (ezr-ondrej) [inline app/presenters/setting_presenter.rb:64]: This should deal with `default = <some_default>` but user set nil (it will be fetched from database and set as explicit `value =` and the `@explicit_value = true` will be set.

**Comment by evgeni** [comment]
It wasn't hammer only tho, UI (Administer > Settings) also showed a lot of "Empty" lines, and some tabs (like bootdisk) didn't render at all and were all white 🤷‍♀️

**Comment by ezr-ondrej** [comment]
> It wasn't hammer only tho, UI (Administer > Settings) also showed a lot of "Empty" lines, and some tabs (like bootdisk) didn't render at all and were all white woman_shrugging

No wonder, we are using the API `settings#index` to get the settings for UI :)
That's why it was proper scary that all tests passed.

EDIT: speaking of `settings#index`: #9070

**Comment by evgeni** [comment]
Is this a can of wasps, or a nest of worms?

So the new patch seems to work for Foreman fine. However, when I try to override the default of a Setting in a theme package, it ignores that override and uses the original default from the Plugin?

Happens with `bootdisk_allowed_types` from `foreman_bootdisk` here. The original default is `["generic", "host", "full_host", "subnet"]`, but I'd like to override it to `["subnet", "full_host"]`, which worked before.

Edit: This seems only to happen for Array defaults, and works for strings?! Or is it the way bootdisk initializes settings? https://github.com/theforeman/foreman_bootdisk/blob/master/app/models/setting/bootdisk.rb

Edit 2: after playing more around with overrides of other things, it's definitely bootdisk!

Edit 3: discovery has the same issue, while templates has not -- I wanted to say this is because they all subclass the `Setting` model, but as it works for `templates`…  But that uses `load_defaults` instead of `default_settings` :woman_shrugging:

**Comment by evgeni** [comment]
After a bit of offline discussion, the theme was overriding things wrongly, and after patching that, it works like it should.

File: app/services/setting_registry.rb [hunk 6]
```diff
@@ -113,7 +113,6 @@ def load
       # Creating missing records as we operate over the DB model while updating the setting
       _new_db_record(definition).save(validate: false)
       definition.updated_at = nil
-      definition.value ||= definition.default
```
Comment (ekohl) [inline app/services/setting_registry.rb:116]: I don't quite understand this bit. Why do you create DB model definitions but don't set a value? Wouldn't it be easier to not save them at all then?

File: app/presenters/setting_presenter.rb [hunk 7]
```diff
@@ -22,6 +22,26 @@ def self.model_name
     Setting.model_name
   end
 
+  # explicit value from mass assignment can not be nil
+  def _assign_attribute(k, v)
```
Comment (ekohl) [inline app/presenters/setting_presenter.rb]: Is this overriding some built in from model?

Comment (evgeni) [reply app/presenters/setting_presenter.rb]: I think it's this: https://github.com/rails/rails/blob/2715b5db76e78620a018269e7cc41330ba6752c0/activemodel/lib/active_model/attribute_assignment.rb#L47-L54

Comment (ekohl) [reply app/presenters/setting_presenter.rb]: Should this method also be private, just like ActiveModel?

File: app/services/setting_registry.rb (hunk 6 above)
Comment (ezr-ondrej) [reply app/services/setting_registry.rb:116]: It would, but I didn't want to go that far :shrug:

File: app/presenters/setting_presenter.rb (hunk 7 above)
Comment (ezr-ondrej) [reply app/presenters/setting_presenter.rb]: Yes, it is. The internal method of ActiveModel, I do not have a better solution for distinguishing between mass assignment and direct assignment.

**Review by ekohl — APPROVED** [review]
@evgeni you verified this, right? If so, should we merge this? Even though we're technically already in stabilization.

**Comment by evgeni** [comment]
> @evgeni you verified this, right? If so, should we merge this? Even though we're technically already in stabilization.

Yes, yes I did.

**Review by evgeni — APPROVED** [review]

**Review by evgeni — APPROVED** [review]
